lib/pbench/server/api/resources/server_configuration.py

```diff
@@ -0,0 +1,152 @@
+from http import HTTPStatus
+from logging import Logger
+
+from flask.json import jsonify
+from flask.wrappers import Request, Response
+
+from pbench.server import PbenchServerConfig
+from pbench.server.api.resources import (
+    API_AUTHORIZATION,
+    API_METHOD,
+    APIAbort,
+    API_OPERATION,
+    ApiBase,
+    ApiParams,
+    ApiSchema,
+    MissingParameters,
+    ParamType,
+    Parameter,
+    Schema,
+)
+from pbench.server.database.models.server_config import (
+    ServerConfig,
+    ServerConfigBadValue,
+    ServerConfigError,
+)
+
+
+class ServerConfiguration(ApiBase):
+    """
+    API class to retrieve and mutate Dataset metadata.
+    """
+
+    def __init__(self, config: PbenchServerConfig, logger: Logger):
+        super().__init__(
+            config,
+            logger,
+            ApiSchema(
+                API_METHOD.PUT,
+                API_OPERATION.UPDATE,
+                uri_schema=Schema(
+                    Parameter("key", ParamType.KEYWORD, keywords=ServerConfig.KEYS)
+                ),
+                # TODO: We aspire to be flexible about how a config option is
+                # specified on a PUT. While the "config" parameter is specifically
+                # intended to support a full JSON body that sets or updates all
+                # options at once, for individual options using the "value"
+                # query parameter is probably easier... but the "value" body
+                # parameter also allows specifying the value in a body.
+                #
+                #   PUT /server/configuration/dataset-lifetime?value=2y
+                #   PUT /server/configuration/dataset-lifetime
+                #       {"value": "2y"}
+                #   PUT /server/configuration/dataset-lifetime ???
+                #       {"config": {"dataset-lifetime": "2y"}}
+                #
+                # Some options may not make sense: need to think about this.
+                query_schema=Schema(Parameter("value", ParamType.STRING)),
+                body_schema=Schema(
+                    Parameter("config", ParamType.JSON, keywords=ServerConfig.KEYS),
+                    Parameter("value", ParamType.JSON),
+                ),
+                authorization=API_AUTHORIZATION.ADMIN,
+            ),
+            ApiSchema(
+                API_METHOD.GET,
+                API_OPERATION.READ,
+                uri_schema=Schema(
+                    Parameter("key", ParamType.KEYWORD, keywords=ServerConfig.KEYS)
+                ),
+                authorization=API_AUTHORIZATION.NONE,
+            ),
```

Comment:
> Flask won't route the trailing slash.

What does that mean?...the request gets sent back to the requestor before it reaches our code?

Reply:
Correct. But I just tried experimenting with a *triple* Flask mapping so that with or without the trailing `/` (and without a key name) it'll return all the config parameters. Maybe I like it?

Reply:
Sounds promising.

Reply:
> Sounds promising.

I played around with this triple-mapping a bit. It works in the config unit tests, but causes a bit of pain in the endpoint configuration test because the trailing slash (without the keyword) generates a second Flask "rule". The normal double mapping with no slash and with `/<key>` are absorbed by the normal mapping logic and don't cause a problem, but the third mapping means I need to write additional (and somewhat hacky) logic in `endpoint_configure.py` to collapse them, and I'm kinda thinking the extra mapping wasn't worth that mess.

That is, if you want to target one key, `/server/configuration/key` and if you want them all, `/server/configuration` ... `/server/configuration/` will get a Flask mapping error (404) just like all the other kids, and we don't worry about it... ?

Reply:
> if you want to target one key, [specify] `/server/configuration/key`

Seems reasonable.

> if you want them all, [specify] `/server/configuration`

Again, seems reasonable.

> [specifying] `/server/configuration/` will get a Flask mapping error (404)

OK, I can live with that.

But, the question is, does this mean that `/server/configuration/key` and `/server/configuration` can, from the user perspective, be considered to be two separate methods, with different requirements for parameters?  And, that notwithstanding, can/would they be handled by a common function which can tell them apart by the missing/empty `key` URI parameter?  If so, I think all's good, here.

Reply:
One of them looks sorta like
```
def get_server_configuration(body: dict):
```
while the other looks like
```
def get_server_configuration_key(key: str):
```
If you pass the parameter to the wrong one, it won't work at all...which makes me think that they are different interfaces (saying nothing about their implementations).